src/ui/default_viewer_setup.ts

```diff
@@ -14,26 +14,124 @@
  * limitations under the License.
  */
 
+import type { UserLayer, UserLayerConstructor } from "#src/layer/index.js";
+import { layerTypes } from "#src/layer/index.js";
 import { StatusMessage } from "#src/status.js";
 import {
   bindDefaultCopyHandler,
   bindDefaultPasteHandler,
 } from "#src/ui/default_clipboard_handling.js";
 import { setDefaultInputEventBindings } from "#src/ui/default_input_event_bindings.js";
 import { makeDefaultViewer } from "#src/ui/default_viewer.js";
-import type { MinimalViewerOptions } from "#src/ui/minimal_viewer.js";
 import { bindTitle } from "#src/ui/title.js";
+import type { Tool } from "#src/ui/tool.js";
+import { restoreTool } from "#src/ui/tool.js";
 import { UrlHashBinding } from "#src/ui/url_hash_binding.js";
+import {
+  verifyObject,
+  verifyObjectProperty,
+  verifyString,
+} from "#src/util/json.js";
 
 declare let NEUROGLANCER_DEFAULT_STATE_FRAGMENT: string | undefined;
 
+type CustomToolBinding = {
+  layer: string;
+  tool: unknown;
+  provider?: string;
+};
+
+type CustomBindings = {
+  [key: string]: CustomToolBinding | string | boolean;
+};
+
+declare const CUSTOM_BINDINGS: CustomBindings | undefined;
+export const hasCustomBindings =
+  typeof CUSTOM_BINDINGS !== "undefined" &&
+  Object.keys(CUSTOM_BINDINGS).length > 0;
+
 /**
  * Sets up the default neuroglancer viewer.
  */
-export function setupDefaultViewer(options?: Partial<MinimalViewerOptions>) {
-  const viewer = ((<any>window).viewer = makeDefaultViewer(options));
+export function setupDefaultViewer() {
+  const viewer = ((<any>window).viewer = makeDefaultViewer());
   setDefaultInputEventBindings(viewer.inputEventBindings);
 
+  const bindNonLayerSpecificTool = (
+    obj: unknown,
+    toolKey: string,
+    desiredLayerType: UserLayerConstructor,
+    desiredProvider?: string,
+  ) => {
+    let previousTool: Tool<object> | undefined;
+    let previousLayer: UserLayer | undefined;
+    if (typeof obj === "string") {
+      obj = { type: obj };
+    }
+    verifyObject(obj);
+    const type = verifyObjectProperty(obj, "type", verifyString);
+    viewer.bindAction(`tool-${type}`, () => {
+      const acceptableLayers = viewer.layerManager.managedLayers.filter(
+        (managedLayer) => {
+          const correctLayerType =
+            managedLayer.layer instanceof desiredLayerType;
+          if (desiredProvider && correctLayerType) {
+            for (const dataSource of managedLayer.layer?.dataSources || []) {
+              const protocol = viewer.dataSourceProvider.getProvider(
+                dataSource.spec.url,
+              )[2];
+              if (protocol === desiredProvider) {
```

Comment:
With the kvstore changes getProvider is not exactly the right method anymore.  Perhaps just have a regexp that matches on the URL?

Reply:
I switched it to use the schemePattern from `"#src/kvstore/url.js";` does that seem fine?